roles/st2/tasks/main.yml

```diff
@@ -48,7 +48,7 @@
 
 - name: Configure StackStorm authentication
   include: auth.yml
-  when: "{{ st2_config.auth.enable|default(st2_auth_enable) if st2_config.auth is defined else st2_auth_enable }}"
+  when: _st2_config_auth_enable|default(st2_auth_enable, true)|bool
```

Comment:
Not sure if `default(st2_auth_enable, true)` is desired, comparing to previous `default(st2_auth_enable)`.

Also, talking about the refactoring, here is a shorter example:
```
when: (st2_config.auth|default({})).enable|default(st2_auth_enable)
```
so we don't create additional `_st2_config_auth_enable` in vars and keep it not that long & minimal as possible.

Reply:
Note: would be nice to include a small code comment about this `when` block.

Reply:
I tried a lot of different things to fix this. That big ugly `_st2_config_auth_enable` is the way I found to fix it. The ternary check that is currently there never could have worked because that's not a proper jinja ternary.

If `when: (st2_config.auth|default({})).enable|default(st2_auth_enable)` works, then perfect, let's use that.

In any case, I was trying to cleanup the output so I didn't have so many errors about "{{ }}" on a when block... It turns out the entire when conditional was not correct, but that's what I'm trying to cleanup.

Reply:
```
ansible all -i "localhost," -m debug -a 'msg="{{ st2_config.auth|default({}).enable|default(st2_auth_enable) }}"' --extra-vars='{"st2_config":{},"st2_auth_enable":'asdf'}'

Reply:
```ansible all -i "localhost," -m debug -a 'msg="{{ st2_config.auth|default({}).enable|default(st2_auth_enable) }}"' --extra-vars='{"st2_config":{},"st2_auth_enable":'asdf'}'

Reply:
This when condition does not work.
```
$ ansible all -i "localhost," -m debug -a 'msg="{{ st2_config.auth|default({}).enable|default(st2_auth_enable) }}"' --extra-vars='{"st2_config":{},"st2_auth_enable":'asdf'}'
localhost | FAILED! => {
    "failed": true, 
    "msg": "template error while templating string: expected token 'end of print statement', got '.'. String: {{ st2_config.auth|default({}).enable|default(st2_auth_enable) }}"
}

$ ansible all -i "localhost," -m debug -a 'msg="{{ st2_config.auth|default({})['enable']|default(st2_auth_enable) }}"' --extra-vars='{"st2_config":{},"st2_auth_enable":'asdf'}'
localhost | FAILED! => {
    "failed": true, 
    "msg": "template error while templating string: expected token 'end of print statement', got '['. String: {{ st2_config.auth|default({})[enable]|default(st2_auth_enable) }}"
}
```

Reply:
This works for the empty `{}` case.
```
ansible all -i "localhost," -m debug -a $'msg="{{ st2_config.auth|default({})|attr(\'enable\')|default(st2_auth_enable) }}"' --extra-vars='{"st2_config":{},"st2_auth_enable":'asdf'}' 
localhost | SUCCESS => {
    "msg": "asdf"
}
```
but not if it's populated:
```
ansible all -i "localhost," -m debug -a $'msg="{{ st2_config.auth|default({})|attr(\'enable\')|default(st2_auth_enable) }}"' --extra-vars='{"st2_config":{"auth":{"enable":"qwer"}},"st2_auth_enable":'asdf'}'
localhost | SUCCESS => {
    "msg": "asdf"
}
```

That's because we need to get an item from the dictionary not an attribute. I don't see a filter for getting something out of a dictionary returned by another filter.

Reply:
Aargh. I missed you're extra parentheses. This works (just like you had it):
```
$ ansible all -i "localhost," -m debug -a 'msg="{{ (st2_config.auth|default({})).enable|default(st2_auth_enable) }}"' --extra-vars='{"st2_config":{},"st2_auth_enable":'asdf'}'
localhost | SUCCESS => {
    "msg": "asdf"
}

$ ansible all -i "localhost," -m debug -a 'msg="{{ (st2_config.auth|default({})).enable|default(st2_auth_enable) }}"' --extra-vars='{"st2_config":{"auth":{"enable":"qwer"}},"st2_auth_enable":'asdf'}'
localhost | SUCCESS => {
    "msg": "qwer"
}
```

OK. I will add this in a few minutes.